### PR TITLE
Implement rendering optimization in the Scheduler Track

### PR DESCRIPTION
**Pull request by karupayun**
In this PR we are finally using the new rendering optimization in the SchedulerTrack
(as it was planned in http://go/stadia-orbit-timer-rendering-improvements).

We are just implementing for SchedulerTrack to measure and have conclusions, but
in the future we will probably have this optimization by default and only have the
old method for GpuSubmissionTrack (unordered timers - http://b/220680172 and a
different implementation who uses depth in a different way - http://b/221024788)
and GpuDebugMarkers (where timers can overlap and the optimization should be
made taking into account both start and end time - http://b/200692451).

This default implementation is very similar to the one in ThreadTracks (and actually,
they might be merged in the future - http://b/205032296).

The implementation is quite-straightforward but I needed to add a hack because a
precision bug I found in GetWorldFromTick.

I briefly benchmark the optimization locally using the 5-min capture in Infiltrator
without Instrumentation (from here:
https://drive.google.com/file/d/139OouMrL6_gy2hYt3x5mhGo2AcIlAL-S/view?usp=sharing).

* 8 Cores, 614k timers per core in average.
* Zoom-out capture entirely only with the Scheduler Track.
  * Before:
     Time to update primitives in the Scheduler: ~133.2375 ms
  * Now:
     Time to update primitives in the Scheduler (1440p screen): ~17.8464 ms  (7.46x)
     Time to update primitives in the Scheduler (1080p screen): ~15.1746 ms

I also measured using a 1.5 minute capture in cumbia. There are more events in the
Scheduler, so rendering was taking ~229ms and now it takes ~14.9ms (1080p)

Tests:
* Load different captures. See there is a timer in each pixel when zooming-out a lot.
* Zoom-in. Test performance didn't decrease.
* Take a capture, see everything works as expected.
Bug: http://b/242971304

### CODE REVIEW

File: src/OrbitGl/SchedulerTrack.cpp [hunk 1]
```diff
@@ -51,12 +52,58 @@ float SchedulerTrack::GetHeight() const {
...
+  const internal::DrawData draw_data =
+      GetDrawData(min_tick, max_tick, GetPos()[0], GetWidth(), &primitive_assembler, timeline_info_,
+                  viewport_, IsCollapsed(), app_->selected_timer(), app_->GetScopeIdToHighlight(),
+                  app_->GetGroupIdToHighlight(), app_->GetHistogramSelectionRange());
+
+  uint32_t resolution_in_pixels = viewport_->WorldToScreen({GetWidth(), 0})[0];
+  float box_height = GetDefaultBoxHeight();
+
+  for (uint32_t depth = 0; depth < GetDepth(); depth++) {
+    float world_timer_y = GetYFromDepth(depth);
```
Comment (reichlfl) [inline src/OrbitGl/SchedulerTrack.cpp:91]: Nit: const float

**Review by reichlfl — COMMENTED** [review]
Lgtm, thanks a lot for this. A suggestion, feel free to ignore or do this in a follow-up PR: Depending on how soon you want to merge the different implementations, would it make sense to leave `TimerTrack::DoUpdatePrimitives` empty and instead provide something like `TimerTrack::RenderTimersUnoptimized`? Then you could call this in all timer track subclasses that do not implement any special optimization and it would be easier to follow which class uses which rendering code instead of omitting the `super` calls.

File: src/OrbitGl/SchedulerTrack.cpp (hunk 1 above)
Comment (florian-kuebler) [reply src/OrbitGl/SchedulerTrack.cpp:91]: If you decide to make the primitives const (which I would appreciate), then also consider "resolution_in_pixels" and "box_height".

**Review by florian-kuebler — APPROVED** [review]
Thanks. I agree with what @reichlfl said, but keep this mostly up to you!

**Review by reichlfl — APPROVED** [review]
I actually meant to approve this.

**Comment by pierricgimmig** [comment]
Thanks for this Pablo!

**Review by pierricgimmig — APPROVED** [review]

**Comment by karupayun** [comment]
> Lgtm, thanks a lot for this. A suggestion, feel free to ignore or do this in a follow-up PR: Depending on how soon you want to merge the different implementations, would it make sense to leave `TimerTrack::DoUpdatePrimitives` empty and instead provide something like `TimerTrack::RenderTimersUnoptimized`? Then you could call this in all timer track subclasses that do not implement any special optimization and it would be easier to follow which class uses which rendering code instead of omitting the `super` calls.

I will include this comment in the tracking bug. This was kind of an experiment in SchedulerTrack, but we need to provide two methods to render timers, I agree.

I'm not sure if it's not better to have the logic directly in TimerTracks for now. TimerTrack::DoUpdatePrimitives could use internal methods (like UpdateOrderedPrimitives or UpdatePrimitivesUnoptimized) depending on the type of the Track, but maybe you are right and this can be chosen in the subclasses and there it can be better explained why we are "not optimizing". 

Anyway, now we have Scheduler as the exception and I prefer this than having 6 different exceptions. As soon as this is implemented, we can solve this tech-debt. For now, I think that having Scheduler as the exception is the best way.

**Pull request merged**